mesonbuild/dependencies/base.py

```diff
@@ -879,6 +875,14 @@ def _search_libs(self, out, out_raw, out_all):
             # Order of -L flags doesn't matter with ld, but it might with other
             # linkers such as MSVC, so prepend them.
             link_args = ['-L' + lp for lp in prefix_libpaths] + link_args
+
+        # Finally, append -Wl,-rpath-link for secondary dependencies
+        if self.clib_compiler and self.clib_compiler.linker and isinstance(self.clib_compiler.linker, GnuLikeDynamicLinkerMixin):
+            # Finding libraries in secondary_libpaths should
+            # have the lowest priority.
+            # https://github.com/mesonbuild/meson/issues/6027
+            link_args += ['-Wl,-rpath-link,' + p for p in secondary_libpaths]
```

Comment:
I found you put `prefix_libpaths` here in the first version of the patch but removed it in later versions. This quickly leads to build failure:
```
The Meson build system
Version: 0.52.0
Source dir: /home/lantw44/gnome/source/baobab
Build dir: /home/lantw44/gnome/build/baobab
Build type: native build
Project name: baobab
Project version: 3.34.0
C compiler for the host machine: clang -B/home/lantw44/.local/bin (clang 8.0.1 "FreeBSD clang version 8.0.1 (tags/RELEASE_801/final 366581) (based on LLVM 8.0.1)")
C linker for the host machine: lld 8.0.1
Vala compiler for the host machine: valac (valac 0.46.1.63-d9c36)
Host machine cpu family: x86_64
Host machine cpu: x86_64
DEPRECATION: Project targetting '>= 0.41.0' but tried to use feature deprecated since '0.48.0': python3 module
Found pkg-config: /usr/local/bin/pkg-config (1.6.3)
Run-time dependency glib-2.0 found: YES 2.63.0
Run-time dependency gio-2.0 found: YES 2.63.0
Run-time dependency gio-unix-2.0 found: YES 2.63.0
Run-time dependency gobject-2.0 found: YES 2.63.0
Run-time dependency gtk+-3.0 found: YES 3.24.12
Library m found: YES
Configuring config.h using configuration
Program /usr/local/bin/python3 found: YES (/usr/local/bin/python3)
Configuring org.gnome.baobab.service using configuration
Found pkg-config: /usr/local/bin/pkg-config (1.6.3)
Build targets in project: 10
WARNING: Deprecated features used:
 * 0.48.0: {'python3 module'}
Found ninja-1.9.0 at /usr/local/bin/ninja
[1/1] Linking target src/baobab.
FAILED: src/baobab 
clang -B/home/lantw44/.local/bin  -o src/baobab 'src/25a6634@@baobab@exe/meson-generated_.._org.gnome.baobab.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-application.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-cellrenderers.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-chart.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-location-list.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-location.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-ringschart.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-scanner.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-treemap.c.o' 'src/25a6634@@baobab@exe/meson-generated_baobab-window.c.o' 'src/25a6634@@baobab@exe/meson-generated_main.c.o' -L/home/lantw44/gnome/devinstall/lib -L/usr/local/lib -Wl,--as-needed -Wl,--no-undefined -march=corei7 -B/home/lantw44/.local/bin -pipe -g3 -Og -gz -fdebug-macro -B/home/lantw44/.local/bin -Wl,--compress-debug-sections=zlib -fuse-ld=bfd -rdynamic -Wl,--start-group /home/lantw44/gnome/devinstall/lib/libglib-2.0.so -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl /home/lantw44/gnome/devinstall/lib/libgio-2.0.so /home/lantw44/gnome/devinstall/lib/libgobject-2.0.so -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl /home/lantw44/gnome/devinstall/lib/libgtk-3.so /home/lantw44/gnome/devinstall/lib/libgdk-3.so /usr/lib/libz.so /home/lantw44/gnome/devinstall/lib/libpangocairo-1.0.so /home/lantw44/gnome/devinstall/lib/libpango-1.0.so /home/lantw44/gnome/devinstall/lib/libharfbuzz.so /home/lantw44/gnome/devinstall/lib/libatk-1.0.so /home/lantw44/gnome/devinstall/lib/libcairo-gobject.so /home/lantw44/gnome/devinstall/lib/libcairo.so /home/lantw44/gnome/devinstall/lib/libgdk_pixbuf-2.0.so -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl -lm -Wl,--end-group
/usr/local/bin/ld.bfd: /home/lantw44/gnome/devinstall/lib/libpango-1.0.so: undefined reference to `fribidi_get_par_embedding_levels_ex'
/usr/local/bin/ld.bfd: /home/lantw44/gnome/devinstall/lib/libpango-1.0.so: undefined reference to `fribidi_get_bracket'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Reply:
correct, it also failed a bunch of unit tests

Reply:
Are you going to add `prefix_libpaths` back?

Reply:
I manually added `prefix_libpaths` back, and there is a new failure:
```
The Meson build system
Version: 0.52.0
Source dir: /home/lantw44/gnome/source/gnome-color-manager
Build dir: /home/lantw44/gnome/build/gnome-color-manager
Build type: native build
Project name: gnome-color-manager
Project version: 3.32.0
C compiler for the host machine: clang -B/home/lantw44/.local/bin (clang 8.0.1 "FreeBSD clang version 8.0.1 (tags/RELEASE_801/final 366581) (based on LLVM 8.0.1)")
C linker for the host machine: lld 8.0.1
C++ compiler for the host machine: clang++ -B/home/lantw44/.local/bin (clang 8.0.1 "FreeBSD clang version 8.0.1 (tags/RELEASE_801/final 366581) (based on LLVM 8.0.1)")
C++ linker for the host machine: lld 8.0.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Compiler for C supports arguments -fstack-protector-strong: YES (cached)
Compiler for C supports arguments -Waggregate-return: YES (cached)
Compiler for C supports arguments -Warray-bounds: YES (cached)
Compiler for C supports arguments -Wcast-align: YES (cached)
Compiler for C supports arguments -Wclobbered: NO (cached)
Compiler for C supports arguments -Wdeclaration-after-statement: YES (cached)
Compiler for C supports arguments -Wempty-body: YES (cached)
Compiler for C supports arguments -Wextra: YES (cached)
WARNING: Consider using the builtin warning_level option instead of adding warning flags by hand.
Compiler for C supports arguments -Wformat=2: YES (cached)
Compiler for C supports arguments -Wformat-nonliteral: YES (cached)
Compiler for C supports arguments -Wformat-security: YES (cached)
Compiler for C supports arguments -Wformat-signedness: NO (cached)
Compiler for C supports arguments -Wignored-qualifiers: YES (cached)
Compiler for C supports arguments -Wimplicit-function-declaration: YES (cached)
Compiler for C supports arguments -Winit-self: YES (cached)
Compiler for C supports arguments -Winline: YES (cached)
Compiler for C supports arguments -Wmissing-declarations: YES (cached)
Compiler for C supports arguments -Wmissing-format-attribute: YES (cached)
Compiler for C supports arguments -Wmissing-include-dirs: YES (cached)
Compiler for C supports arguments -Wmissing-noreturn: YES (cached)
Compiler for C supports arguments -Wmissing-parameter-type: NO (cached)
Compiler for C supports arguments -Wmissing-prototypes: YES (cached)
Compiler for C supports arguments -Wnested-externs: YES (cached)
Compiler for C supports arguments -Wno-discarded-qualifiers: NO (cached)
Compiler for C supports arguments -Wno-missing-field-initializers: YES (cached)
Compiler for C supports arguments -Wno-strict-aliasing: YES (cached)
Compiler for C supports arguments -Wno-suggest-attribute=format: NO (cached)
Compiler for C supports arguments -Wno-unused-parameter: YES (cached)
Compiler for C supports arguments -Wold-style-definition: YES (cached)
Compiler for C supports arguments -Woverride-init: YES (cached)
Compiler for C supports arguments -Wpacked: YES (cached)
Compiler for C supports arguments -Wpointer-arith: YES (cached)
Compiler for C supports arguments -Wredundant-decls: YES (cached)
Compiler for C supports arguments -Wreturn-type: YES (cached)
Compiler for C supports arguments -Wshadow: YES (cached)
Compiler for C supports arguments -Wsign-compare: YES (cached)
Compiler for C supports arguments -Wstrict-aliasing: YES (cached)
Compiler for C supports arguments -Wstrict-prototypes: YES (cached)
Compiler for C supports arguments -Wswitch-default: YES (cached)
Compiler for C supports arguments -Wtype-limits: YES (cached)
Compiler for C supports arguments -Wundef: YES (cached)
Compiler for C supports arguments -Wuninitialized: YES (cached)
Compiler for C supports arguments -Wunused-but-set-variable: NO (cached)
Compiler for C supports arguments -Wwrite-strings: YES (cached)
Compiler for C supports link arguments -Wl,-z,relro: YES (cached)
Compiler for C supports link arguments -Wl,-z,now: YES (cached)
Found pkg-config: /usr/local/bin/pkg-config (1.6.3)
Run-time dependency gio-2.0 found: YES 2.63.0
Run-time dependency gtk+-3.0 found: YES 3.24.12
Run-time dependency libexif found: YES 0.6.21
Run-time dependency libtiff-4 found: YES 4.0.10
Run-time dependency colord found: YES 1.4.5
Run-time dependency colord-gtk found: YES 0.2.1
Run-time dependency libcanberra-gtk3 found: YES 0.30
Library libm found: YES
Run-time dependency lcms2 found: YES 2.9
Run-time dependency vte-2.91 found: YES 0.59.1
Run-time dependency exiv2 found: YES 0.27.2
Dependency exiv2 found: YES (cached)
Configuring config.h using configuration
Program docbook2man found: YES (/usr/local/bin/docbook2man)
Found pkg-config: /usr/local/bin/pkg-config (1.6.3)
Program meson_post_install.sh found: YES (/bin/sh /home/lantw44/gnome/source/gnome-color-manager/meson_post_install.sh)
Build targets in project: 26
Found ninja-1.9.0 at /usr/local/bin/ninja
[1/20] Linking target src/gcm-inspect.
FAILED: src/gcm-inspect 
clang -B/home/lantw44/.local/bin  -o src/gcm-inspect 'src/25a6634@@gcm-inspect@exe/gcm-inspect.c.o' 'src/25a6634@@gcm-inspect@exe/gcm-cie-widget.c.o' 'src/25a6634@@gcm-inspect@exe/gcm-debug.c.o' 'src/25a6634@@gcm-inspect@exe/gcm-exif.c.o' 'src/25a6634@@gcm-inspect@exe/gcm-print.c.o' 'src/25a6634@@gcm-inspect@exe/gcm-trc-widget.c.o' 'src/25a6634@@gcm-inspect@exe/gcm-utils.c.o' -L/home/lantw44/gnome/devinstall/lib -L/usr/local/lib -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-z,now -march=corei7 -B/home/lantw44/.local/bin -pipe -g3 -Og -gz -fdebug-macro -B/home/lantw44/.local/bin -Wl,--compress-debug-sections=zlib -fuse-ld=bfd -Wl,--start-group /usr/local/lib/libtiff.so -Wl,-rpath-link,/usr/local/lib /home/lantw44/gnome/devinstall/lib/libcolord-gtk.so /home/lantw44/gnome/devinstall/lib/libcolord.so /home/lantw44/gnome/devinstall/lib/libgtk-3.so /home/lantw44/gnome/devinstall/lib/libgdk-3.so /usr/lib/libz.so /home/lantw44/gnome/devinstall/lib/libpangocairo-1.0.so /home/lantw44/gnome/devinstall/lib/libpango-1.0.so /home/lantw44/gnome/devinstall/lib/libharfbuzz.so /home/lantw44/gnome/devinstall/lib/libatk-1.0.so /home/lantw44/gnome/devinstall/lib/libcairo-gobject.so /home/lantw44/gnome/devinstall/lib/libcairo.so /home/lantw44/gnome/devinstall/lib/libgdk_pixbuf-2.0.so /home/lantw44/gnome/devinstall/lib/libgio-2.0.so /home/lantw44/gnome/devinstall/lib/libgobject-2.0.so /home/lantw44/gnome/devinstall/lib/libglib-2.0.so -Wl,-rpath-link,/home/lantw44/gnome/devinstall/lib -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl /usr/lib/libm.so /usr/local/lib/libexif.so -Wl,-rpath-link,/usr/local/lib -Wl,-rpath-link,/home/lantw44/gnome/devinstall/lib -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl -Wl,-rpath-link,/home/lantw44/gnome/devinstall/lib -Wl,-rpath-link,/usr/local/lib -L/home/lantw44/gnome/devinstall/lib -lintl -Wl,--end-group
/usr/local/bin/ld.bfd: /home/lantw44/gnome/devinstall/lib/libpango-1.0.so: undefined reference to `fribidi_get_par_embedding_levels_ex'
/usr/local/bin/ld.bfd: /home/lantw44/gnome/devinstall/lib/libpango-1.0.so: undefined reference to `fribidi_get_bracket'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

So doing point 1 I mentioned in https://github.com/mesonbuild/meson/issues/6027 isn't enough, and the problem described in https://github.com/mesonbuild/meson/issues/6027#issuecomment-541334015 does exist.